File: Umix/emulator/mem.cc
#include "emulator/mem.h"

void Mem::AllocZeroPage(uint32_t size) {
  if (pages_.empty())
    pages_.resize(1);
  ResetPage(0, size);
}

uint32_t Mem::AllocPage(uint32_t size) {
  uint32_t page;
  if (free_pages_.empty())
  {
    page = pages_.size();
    pages_.resize(page+1);
  } else {
    page = free_pages_.back();
    free_pages_.pop_back();
  }
  ResetPage(page, size);
  return page;
}

void Mem::FreePage(uint32_t page) {
  delete[] pages_[page].data;
  pages_[page].data = nullptr;
  free_pages_.push_back(page);
}

void Mem::CopyPageToZeroPage(uint32_t page) {
  ResetPage(0, pages_[page].size);
  uint32_t* src = pages_[page].data;
  uint32_t* dst = pages_[0].data;
  for (uint32_t offset = 0; offset < pages_[page].size; offset++)
    dst[offset] = src[offset];
}

uint32_t Mem::Read(uint32_t page, uint32_t offset) {
  return pages_[page].data[offset];
}

void Mem::Write(uint32_t page, uint32_t offset, uint32_t value) {
  pages_[page].data[offset] = value;
}

void Mem::ResetPage(uint32_t page, uint32_t size) {
  if (pages_[page].data)
    delete[] pages_[page].data;
  pages_[page].data = new uint32_t[size];
  pages_[page].size = size;
  for (uint32_t offset = 0; offset < size; offset++)
    pages_[page].data[offset] = 0;
}

File: Umix/emulator-rs/src/mem.rs
pub struct Mem {
  pages:Vec<Vec<u32>>,
  free_pages:Vec<usize>,
}

impl Mem {
  pub fn new() -> Mem {
    Mem {
      pages:vec![],
      free_pages:vec![]
    }
  }

  pub fn alloc_zero_page(&mut self, size:u32) {
    assert!(self.pages.is_empty());
    let page = self.alloc_page(size);
    assert!(page == 0);
  }

  pub fn alloc_page(&mut self, size:u32) -> u32 {
    if self.free_pages.is_empty() {
      self.pages.push(vec![]);
      self.free_pages.push(self.pages.len() - 1);
    }
    let page = self.free_pages.pop().unwrap();
    self.pages[page].resize(size as usize, 0);
    return page as u32;
  }

  pub fn free_page(&mut self, page:u32) {
    let page = page as usize;
    self.pages[page].clear();
    self.pages[page].shrink_to_fit();
    self.free_pages.push(page);
  }

  pub fn copy_page_to_zero_page(&mut self, page:u32) {
    assert!(!self.pages.is_empty());
    let page = self.pages[page as usize].to_vec();
    self.pages[0] = page;

  }

  pub fn read(&self, page:u32, offset:u32) -> u32 {
    return self.pages[page as usize][offset as usize];
  }

  pub fn write(&mut self, page:u32, offset:u32, value:u32) {
    self.pages[page as usize][offset as usize] = value;
  }
}